numcodecs/shuffle.pyx

```diff
@@ -0,0 +1,94 @@
+# cython: embedsignature=True
+# cython: profile=False
+# cython: linetrace=False
+# cython: binding=False
+# cython: language_level=3
+
+import numpy as np
+from .compat import ensure_contiguous_ndarray
+from .abc import Codec
+
+cimport cython
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void _doShuffle(const unsigned char[::1] src, unsigned char[::1] des, Py_ssize_t element_size) nogil:
+    cdef Py_ssize_t count, i, j, offset, byte_index
+    count = len(src) // element_size
+    for i in range(count):
+        offset = i*element_size
+        for byte_index in range(element_size):
+            j = byte_index*count + i
+            des[j] = src[offset + byte_index]
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef void _doUnshuffle(const unsigned char[::1] src, unsigned char[::1] des, Py_ssize_t element_size) nogil:
+    cdef Py_ssize_t count, i, j, offset, byte_index
+    count = len(src) // element_size
+    for i in range(element_size):
+        offset = i*count
+        for byte_index in range(count):
+            j = byte_index*element_size + i
+            des[j] = src[offset+byte_index]
+
+
+class Shuffle(Codec):
+    """Codec providing shuffle
+
+    Parameters
+    ----------
+    elementsize : int
+        Size in bytes of the array elements.  Default = 4
+
+    """
+
+    codec_id = 'shuffle'
+
+    def __init__(self, elementsize=4):
+        self.elementsize = elementsize
+
+    def encode(self, buf, out=None):
+        buf = ensure_contiguous_ndarray(buf)
```

Comment:
Most of `encode` and `decode` is the same, could make a separate method like `_prepare_array`. This would shorten the code and, I suspect, have no impact on performance. Only a mild suggestion.

Reply:
Thanks, have added _prepare_array suggestion. The flow control seems a little odd with the return, possibly a more elegant way to handle the input checking...